File: 05-harness-io/node.js/src/data/feature-flag-repository.ts
import { Client } from '@harnessio/ff-nodejs-server-sdk';
import { FeatureFlags } from '../types/feature-flag.js';


/*
Harness.io SDK docs
https://developer.harness.io/docs/feature-flags/use-ff/ff-sdks/server-sdks/node-js-sdk-reference
*/

const HARNESS_IO_KEY = process.env.HARNESS_IO_KEY;
if (!HARNESS_IO_KEY) {
  throw new Error('set HARNESS_IO_KEY environment variable');
}
const client = new Client(HARNESS_IO_KEY);


export default async function getFlag(key: FeatureFlags, name?: string | undefined, defaultValue: boolean = false): Promise<boolean> {
  const target = {
    identifier: 'demo',
    name: 'name',
    attributes: {
       name
    }
  };

  const value = await client.boolVariation(key, target, defaultValue);
  return value;
}

export async function getAllFlags(name?: string | undefined): Promise<{ [key: string]: boolean }> {
  // Harness doesn't have a GetAll in the SDK
  // https://apidocs.harness.io/tag/Feature-Flags/#operation/GetAllFeatures

  const results = {
    'ONE': await getFlag(FeatureFlags.ONE, name),
    'TWO': await getFlag(FeatureFlags.TWO, name),
    'THREE': await getFlag(FeatureFlags.THREE, name)
  }
  return results;
}

File: 05-harness-io/dotnet/Data/FeatureFlagRepository.cs
namespace FeatureFlagHarnessio.Data;

/*
Harness.io feature flags docs:
https://developer.harness.io/docs/feature-flags/use-ff/ff-sdks/server-sdks/net-sdk-reference
*/

public interface IFeatureFlagRepository
{
    bool FeatureIsEnabled(FeatureFlags featureFlag, string? user = null, bool defaultValue = false);
    Dictionary<string, bool> GetAllFeatureFlags(string? user = null);
}

public class FeatureFlagRepository(ICfClient harnessioClient) : IFeatureFlagRepository
{
    public bool FeatureIsEnabled(FeatureFlags featureFlag, string? user = null, bool defaultValue = false)
    {
        Target target = Target.builder()
                .Name("name")
                .Identifier(user)
                .build();

        bool flagValue = harnessioClient.boolVariation(featureFlag.ToString(), target, defaultValue);
        return flagValue;
    }

    public Dictionary<string, bool> GetAllFeatureFlags(string? user = null)
    {
        // Harness doesn't have a GetAll in the SDK
        // https://apidocs.harness.io/tag/Feature-Flags/#operation/GetAllFeatures

        var results = new Dictionary<string, bool>();
        foreach (var flag in Enum.GetValues<FeatureFlags>())
        {
            results.Add(flag.ToString(), FeatureIsEnabled(flag));
        }
        return results;
    }
}
